DevDocumentation/ARCHITECTURE.md

```diff
@@ -1,3 +1,224 @@
+- Envoy Proxy - Proxy that handles routing all user traffic (*in-process)
+- Ambassador Agent - provides connectivity between cluster and Ambassador Cloud.  (*own container*)\
+
+## Custom Resource Definitions (CRD)
+
+Kuberenetes allows extending the Kubernetes API through the use of Customer Resource Definitions (aka CRD's) which allow solutions like Emissary-ingress to provide custom abstractions that provide a K8s familiar developer experience through declarative configuration. CRD's provide validation, strong typing, structured data, versioning and are persisted in `etcd` along with the core Kuberenetes resources.
+
+Emissary-ingress provides a set of CRD's that are applied to a cluster and then are watched by Emissary-ingress. Emissary-ingress then uses the data from these CRD's along with the standard K8's resources (services, endpoints, etc...) to dynamically generate Envoy Proxy configuration. Depending on the version of Emissary-ingress there might be multiple versions of the CRD's that are suppported.
+
+You can read the user documentation (see additional reading below) to find out more about all the various CRD's that are used and how to configure them. For understanding, how they are defined you can take a look in `pkg/getambassador.io/*` directory. In this directory, you will find a directory per version of the CRD's and for each version you will see the `Golang` structs that define the data structures that are used for each of the Emissary-ingress custom resources. It's recommended to read the `doc.go` file for information about api guidelines followed and how the comment markers are used by the build system.
+
+The build system (`make`) uses [controller-gen](https://book.kubebuilder.io/reference/controller-gen.html) to generate the required YAML representation for the customer resources that can be found at `pkg/getambassador.io/crds.yaml`. This file is auto-generated and checked into the respository. This is the file that is applied to a cluster extending the Kubernetes API. If any changes are made to the custom resources then it needs to be re-generated and checked-in as part of your PR. Running `make generate` will trigger the generation of this file and other generated files (`protobufs`) that are checked into the respository as well. If you want to see more about the build process take a look at `build-aux/generate.mk`.
+
+> **Annotations**:  K8s allows developers to provide Annotations as well to the standard K8s Resources (Services, Ingress, etc...). Annotations were the preferred method of configuring early versions of Emissary-ingress but annotations did not provide validation and can be error prone. However, with the introduction of CRD's these are now the preferred method and annotations are only supported for backwards compatibility. We won't discuss the annotations much here due to this but rather making you aware that they exist.
```

Comment:
Idk if annotations were preferred but rather the only option available before the advent of CRDs